### PR TITLE
Add p-refinement criterion based on constraints

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Based on Sec. 6.1.4 in Szilagy 2014. This should work with the GH two-index and three-index constraints. When testing this in the wild we may want to add some h-refinement criterion based on constraints, and try to eliminate the coarsening factor.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * If the coarsening factor turns out to be hard to choose, then we can try to
+ * eliminate it by projecting the variables to a lower polynomial order before
+ * computing constraints, or something like that.
+ *
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the constraints to be monitored. These
+ * must be tensors of rank 2 or higher. The first index must be a lower spatial
+ * index that originates from a derivative.
+ */
+template <size_t Dim, typename TensorTags>
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp:114]: Are these going to be evaluated using the `ObservationBox`? I don't want to store the 2 & 4-index constraints in the DataBox. We don't have enough RAM for that.

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  Constraints(std::vector<std::string> vars_to_monitor, double abs_target,
+              double coarsening_factor, const Options::Context& context = {});
+
+  /// \cond
+  explicit Constraints(CkMigrateMessage* msg);
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Constraints);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: Set to `TensorTags`

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  std::array<Flag, Dim> operator()(const db::DataBox<DbTagsList>& box,
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: `DataBox` -> `ObservationBox`

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+  tmpl::for_each<TensorTags>([&result, &box, &jacobian, &logical_constraints,
+                              &buffer, this](const auto tag_v) {
+    // Stop if we have already decided to refine every dimension
+    if (result == make_array<Dim>(Flag::IncreaseResolution)) {
+      return;
+    }
+    using tag = tmpl::type_from<std::decay_t<decltype(tag_v)>>;
+    const std::string tag_name = db::tag_name<tag>();
+    // Skip if this tensor is not being monitored
+    if (alg::find(vars_to_monitor_, tag_name) == vars_to_monitor_.end()) {
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: `if (not alg::found)`

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+template <typename DbTagsList, typename Metavariables>
+std::array<Flag, Dim> Constraints<Dim, TensorTags>::operator()(
+    const db::DataBox<DbTagsList>& box,
+    Parallel::GlobalCache<Metavariables>& /*cache*/,
+    const ElementId<Dim>& /*element_id*/) const {
+  auto result = make_array<Dim>(Flag::Undefined);
+  const auto& jacobian = db::get<
+      domain::Tags::Jacobian<Dim, Frame::ElementLogical, Frame::Inertial>>(box);
+  tnsr::i<DataVector, Dim, Frame::ElementLogical> logical_constraints{};
+  DataVector buffer{};
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: set the size correctly here so it doesn't get resized later

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+template <size_t Dim, typename TensorTags>
+template <typename DbTagsList, typename Metavariables>
+std::array<Flag, Dim> Constraints<Dim, TensorTags>::operator()(
+    const db::DataBox<DbTagsList>& box,
+    Parallel::GlobalCache<Metavariables>& /*cache*/,
+    const ElementId<Dim>& /*element_id*/) const {
+  auto result = make_array<Dim>(Flag::Undefined);
+  const auto& jacobian = db::get<
+      domain::Tags::Jacobian<Dim, Frame::ElementLogical, Frame::Inertial>>(box);
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: Please make this work with FD by using the `Events` version of this tag

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  // constraint is above the target, the element will increase p refinement in
+  // that dimension. And only if all constraints are below the coarsening
+  // threshold will the element decrease p refinement in that dimension.
+  for (size_t d = 0; d < Dim; ++d) {
+    // Skip this dimension if we have already decided to refine it
+    if (gsl::at(*result, d) == Flag::IncreaseResolution) {
+      continue;
+    }
+    const double constraint_norm = blaze::l2Norm(logical_constraints.get(d)) /
+                                   sqrt(logical_constraints.begin()->size());
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.cpp]: precompute the sqrt part (sqrt is super expensive) outside of loop

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        *buffer += jacobian.get(i, k) * constraints_tensor.get(full_index);
+      }
+      result->get(k) += square(*buffer);
+    }
+    // Compute normalization factor
+    // Possible performance optimization: unroll first iteration of loop to
+    // avoid zeroing buffer
+    *buffer = 0.;
+    for (size_t i = 0; i < Dim; ++i) {
+      *buffer += square(jacobian.get(i, k));
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.cpp]: Can you precompute this for _all_ constraints once?

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+namespace amr::Criteria {
+
+namespace Constraints_detail {
+
+/*!
+ * \brief Computes a measure of the constraints in each logical direction of the
+ * grid
+ *
+ * \requires `constraints_tensor` to be a tensor of rank 2 or higher. The first
+ * index must be a lower spatial index that originates from a derivative.
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp:54]: add a static_assert to enforce?

File: src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+ *
+ * We follow \cite Szilagyi2014fna, Eq. (62)-(64) to compute:
+ *
+ * \begin{align}
+ * \Epsilon_\hat{k} = \frac{1}{N_\hat{k}} \sqrt{\sum_{a\ldots} \left(\sum_{i}
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp]: &= 2x to align equations

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
sorry for being a bit brief, wanted to get this to you finally...

**Comment by nilsvu** [comment]
Fixup pushed

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash!

**Comment by nilsvu** [comment]
Done

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsvu** [comment]
@kidder do you want to take a look at this before we merge it?

**Pull request merged**